code/modules/reagents/reagent_containers/inhaler.dm

```diff
@@ -0,0 +1,288 @@
+/obj/item/inhaler/proc/try_insert_canister(obj/item/reagent_containers/inhaler_canister/new_canister, mob/living/user, params)
+	if (!isnull(canister))
+		balloon_alert(user, "remove the existing canister!")
+		return FALSE
+
+	balloon_alert(user, "inserting canister...")
+	playsound(src, new_canister.pre_insert_sound, new_canister.pre_insert_volume)
+	if (!do_after(user, new_canister.insertion_time, src))
+		return FALSE
+	playsound(src, new_canister.post_insert_sound, new_canister.post_insert_volume)
+	balloon_alert(user, "canister inserted")
+	set_canister(new_canister, user)
+
+	return TRUE
+
+/// Setter proc for [canister]. Moves the existing canister out of the inhaler, while moving a new canister inside and registering it.
+/obj/item/inhaler/proc/set_canister(obj/item/reagent_containers/inhaler_canister/new_canister, mob/living/user)
+	if (!isnull(canister))
+		if (iscarbon(loc))
+			var/mob/living/carbon/carbon_loc = loc
+			INVOKE_ASYNC(carbon_loc, TYPE_PROC_REF(/mob/living/carbon, put_in_hands), canister)
+		else if (!isnull(loc))
+			canister.forceMove(loc)
+		UnregisterSignal(canister, COMSIG_QDELETING)
+
+	canister = new_canister
+	canister?.forceMove(src)
+	RegisterSignal(canister, COMSIG_QDELETING, PROC_REF(canister_deleting))
+	update_canister_underlay()
+
+/// Updates our underlays with our canister, if any.
+/obj/item/inhaler/proc/update_canister_underlay()
+	if (isnull(canister))
+		underlays -= canister_underlay
+		canister_underlay = null
+	else if (isnull(canister_underlay))
+		canister_underlay = mutable_appearance(canister.icon, canister.icon_state)
+		canister_underlay.pixel_y = canister_underlay_y_offset
+		underlays += canister_underlay
+
+/// Determines if we can be used. Fails on no canister, empty canister, invalid targets, or non-breathing targets.
+/obj/item/inhaler/proc/can_puff(mob/living/target_mob, mob/living/user, silent = FALSE)
+	if (isnull(canister))
+		if (!silent)
+			balloon_alert(user, "no canister!")
+		return FALSE
+	if (isnull(canister.reagents) || canister.reagents.total_volume <= 0)
+		if (!silent)
+			balloon_alert(user, "canister is empty!")
+		return FALSE
+	if (!iscarbon(target_mob)) // maybe mix this into a general has mouth check
+		if (!silent)
+			balloon_alert(user, "not a carbon!")
+		return FALSE
+	if (target_mob.is_mouth_covered())
+		if (!silent)
+			balloon_alert(user, "expose the mouth!")
+		return FALSE
+	if (!target_mob.can_breathe_reagents())
+		if (!silent)
+			balloon_alert(user, "not breathing!")
+		return FALSE
+
+	return TRUE
+
+/obj/item/inhaler/proc/canister_deleting(datum/signal_source)
+	SIGNAL_HANDLER
+
+	set_canister(null)
+
+/obj/item/reagent_containers/inhaler_canister
+	name = "inhaler canister"
+	desc = "A small canister filled with aerosolized reagents for use in a inhaler."
+	w_class = WEIGHT_CLASS_TINY
+
+	icon = 'icons/obj/medical/chemical.dmi'
+	icon_state = "inhaler_canister"
+
+	amount_per_transfer_from_this = 5
+	volume = 25
```

Comment:
This can be removed because the default volume for a reagent container is 30 units. This will also fix your unit tests